File: app/src/main/java/shts/jp/android/nogifeed/api/AsyncNewsClient.java
package shts.jp.android.nogifeed.api;

import android.content.Context;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.io.IOException;

import rx.Observable;
import rx.Subscriber;
import shts.jp.android.nogifeed.entities.News;
import shts.jp.android.nogifeed.entities.NewsList;

public class AsyncNewsClient {

    private static final String TAG = AsyncNewsClient.class.getSimpleName();
    private static final String URL_NEWS = "http://www.nogizaka46.com/smph/news/";

    public static Observable<NewsList> get(@NonNull final Context context) {
        return Observable.create(new Observable.OnSubscribe<NewsList>() {
            @Override
            public void call(Subscriber<? super NewsList> subscriber) {
                try {
                    NewsList newsList = getNewsFeed(URL_NEWS);
                    newsList.filter(context);
                    newsList.sort();
                    subscriber.onNext(newsList);
                    subscriber.onCompleted();
                } catch (Throwable e) {
                    subscriber.onError(e);
                }
            }
        });
    }

    private static NewsList getNewsFeed(final String url) throws IOException {
        final NewsList newsList = new NewsList();

        Document document = Jsoup.connect(url).get();
        Element body = document.body();
        Element dl = body.getElementsByTag("dl").get(0);

        final int length = dl.getElementsByTag("dt").size();
        for (int i = 0; i < length; i++) {
            String iconTypeText = dl.getElementsByTag("dt").get(i).className();
            if (TextUtils.isEmpty(iconTypeText)) {
                iconTypeText = "icon4";
            }
            final String date = dl.getElementsByTag("dt").get(i).text();
            final String title = dl.getElementsByTag("dd").get(i).text();
            final String newsUrl = dl.getElementsByTag("a").get(i).attr("href");
            newsList.add(new News(date, iconTypeText, newsUrl, title));
        }
        return newsList;
    }
}

File: app/src/main/java/shts/jp/android/nogifeed/api/AsyncNewsClient.kt
package shts.jp.android.nogifeed.api

import android.content.Context
import android.text.TextUtils
import org.jsoup.Jsoup
import rx.Observable
import shts.jp.android.nogifeed.entities.News
import shts.jp.android.nogifeed.entities.NewsList
import java.io.IOException

class AsyncNewsClient {

    companion object {

        private val URL_NEWS = "http://www.nogizaka46.com/smph/news/"

        fun get(context: Context): Observable<NewsList> {
            return Observable.create { subscriber ->
                try {
                    val newsList = getNewsFeed(URL_NEWS)
                    newsList.filter(context)
                    newsList.sort()
                    subscriber.onNext(newsList)
                    subscriber.onCompleted()
                } catch (e: Throwable) {
                    subscriber.onError(e)
                }
            }
        }

        @Throws(IOException::class)
        private fun getNewsFeed(url: String): NewsList {
            val newsList = NewsList()

            val document = Jsoup.connect(url).get()
            val body = document.body()
            val dl = body.getElementsByTag("dl")[0]

            val length = dl.getElementsByTag("dt").size
            for (i in 0 until length) {
                var iconTypeText = dl.getElementsByTag("dt")[i].className()
                if (TextUtils.isEmpty(iconTypeText)) {
                    iconTypeText = "icon4"
                }
                val date = dl.getElementsByTag("dt")[i].text()
                val title = dl.getElementsByTag("dd")[i].text()
                val newsUrl = dl.getElementsByTag("a")[i].attr("href")
                newsList.add(News(date, iconTypeText, newsUrl, title))
            }
            return newsList
        }
    }
}
